src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -906,46 +905,108 @@ void MergeTreeData::loadDataParts(bool skip_sanity_checks)
+                    MergeTreeWriteAheadLog wal(*this, disk_ptr, it->name());
+                    for (auto && part : wal.restore(metadata_snapshot, getContext()))
+                        disk_wal_parts.push_back(std::move(part));
+                }
             }
-        }
+        });
+    }
+
+    pool.wait();
+
+    for (auto & [_, disk_wal_parts] : disk_wal_part_map)
+        parts_from_wal.insert(
+            parts_from_wal.end(), std::make_move_iterator(disk_wal_parts.begin()), std::make_move_iterator(disk_wal_parts.end()));
+
+    size_t num_parts = 0;
+    std::queue<std::vector<std::pair<String, DiskPtr>>> parts_queue;
+    for (auto & [_, disk_parts] : disk_part_map)
+    {
+        num_parts += disk_parts.size();
+        parts_queue.push(std::move(disk_parts));
     }
 
+
     auto part_lock = lockParts();
     data_parts_indexes.clear();
 
-    if (part_names_with_disks.empty() && parts_from_wal.empty())
+    if (num_parts == 0 && parts_from_wal.empty())
     {
         LOG_DEBUG(log, "There are no data parts");
         return;
     }
 
-    /// Parallel loading of data parts.
-    size_t num_threads = std::min(size_t(settings->max_part_loading_threads), part_names_with_disks.size());
+    pool.setMaxThreads(std::min(size_t(settings->max_part_loading_threads), num_parts));
+    size_t num_threads = pool.getMaxThreads();
+    size_t min_parts_per_thread = (num_parts - 1) / num_threads + 1;
+
+    /// Prepare data parts for parallel loading. Threads will focus on given disk first, then steal
+    /// others' tasks when finish current disk part loading process.
+    std::vector<std::vector<std::pair<String, DiskPtr>>> threads_parts(num_threads);
+    std::set<size_t> remaining_thread_parts;
+    for (size_t i = 0; i < num_threads && !parts_queue.empty(); ++i)
+    {
+        remaining_thread_parts.insert(i);
+        auto need_parts = min_parts_per_thread;
+        auto & thread_parts = threads_parts[i];
+
+        while (need_parts > 0 && !parts_queue.empty())
+        {
+            auto & current_parts = parts_queue.front();
+            if (current_parts.size() <= need_parts)
+            {
+                thread_parts.insert(thread_parts.end(), current_parts.begin(), current_parts.end());
+                parts_queue.pop();
+                need_parts -= current_parts.size();
+            }
+            else
+            {
+                thread_parts.insert(thread_parts.end(), current_parts.end() - need_parts, current_parts.end());
+                current_parts.resize(current_parts.size() - need_parts);
+                need_parts = 0;
+            }
+        }
+
+        /// Before processing next thread, change disk if possible.
+        /// Different threads will likely start loading parts from different disk,
+        /// which may improve read parallelism for JBOD.
+        if (parts_queue.size() > 1)
+        {
+            parts_queue.push(std::move(parts_queue.front()));
+            parts_queue.pop();
+        }
```

Comment:
But why these is required, does order of disks makes any difference?

Reply:
> does order of disks makes any difference?

Yes, ideally we'd like to load parts from different disks in parallel. Current improvement is 24 min -> 16 min.

Reply:
But this PR already does loading parts in parallel fro different disks, AFAIU this line is only to change the order of those disks, no?

Reply:
Different threads will first load parts from dedicated disks, only after that will try to "steal" parts from other disks.

Reply:
>Different threads will first load parts from dedicated disks, only after that will try to "steal" parts from other disks.

I'm not sure that I completely understand.

Let's assume you have the following parts distribution:
- disk1: 2 parts
- disk2: 2 parts
- disk3: 12 parts

And:
- `max_part_loading_threads=3`
- hence `min_parts_per_thread=6`

And according to the `while (need_parts > 0 && !parts_queue.empty())` lop:
- thread1: will load disk1 and disk2 and 2 parts from disk3
- thread2: will load 6 parts from disk3
- thread3: will load 4 parts from disk3

Am I understand correctly?
If so this does not looks right.

*P.S. it seems that this can be checked with `.sh` test?*

Reply:
Yes, in such skewed part distribution it can be improved (though I doubt the effectiveness). In general we will have much more parts than the number of threads or else there is no need to speed up.

> P.S. it seems that this can be checked with .sh test?

Not sure. I'll add some asserts at least.

Reply:
Logs contains thread ids, so you can check logs after DETACH/ATTACH, although not sure that it worth it...